library/ecjpake.c

```diff
@@ -1107,6 +1112,7 @@ int mbedtls_ecjpake_self_test( int verbose )
 
     if( verbose != 0 )
         mbedtls_printf( "passed\n" );
+#endif /* ! MBEDTLS_ECJPAKE_ALT */
 
 cleanup:
     mbedtls_ecjpake_free( &cli );
```

Comment:
Please bear with me as my knowledge on ECJPAKE and alternative implementations is limited. Also feel free to correct me if I've gotten anything here wrong.

Can you please explain why alternative ECJPAKE implementations cannot use the same test vectors or how consuming entropy differently would make the test vectors invalid.

To me it feels too relaxed to just skip (almost) all of the tests if an alternative implementation is used. Surely some alternative implementations may still want to use the same self-tests? If that is the case then is there some way the changes could be limited to only those implementations that actively enable or disable them, perhaps reusing the existing `MBEDTLS_SELF_TEST`?

Reply:
Sure.

We're not skipping almost all of the tests. What we're skipping is the tests which rely on a reference handshake. The reason those are problematic for use with external implementations (typically accelerators), is that those rely on knowing how the random number is chosen during the handshake procedure.

Alternative implementations may draw entropy from the entropy function in a different way vs the mbed TLS implementation, may construct the random number a different way from the drawn entropy, or even rely on a different (hardware) entropy source entirely. This leads to not being able to use the test, since the generated handshake would extremely likely be different from what the mbed TLS implementation would construct out of the known entropy.

Reply:
Thank you for the detailed explanation.

As shown by my previous comment https://github.com/ARMmbed/mbedtls/pull/4007#discussion_r560335805 somehow I missed that there were two tests and that this is only disabling one of them. In that case I think these changes look sensible and would be useful to add.